lib/mayaUsd/ufe/UsdShaderAttributeDef.cpp

```diff
@@ -170,6 +170,13 @@ static const MetadataMap _metaMap = {
 Ufe::Value UsdShaderAttributeDef::getMetadata(const std::string& key) const
 {
     TF_DEV_AXIOM(fShaderAttributeDef);
+
+#if (UFE_PREVIEW_VERSION_NUM >= 6000)
+    if (key == Ufe::AttributeDef::kNativeType) {
+        return Ufe::Value(fShaderAttributeDef->GetType().GetString());
+    }
```

Comment:
This is the Sdr type, which is OK but not perfect.
To compare, run this in the Maya script editor:
```python
from pxr import Sdr
r = Sdr.Registry()
n = r.GetNodeByIdentifier("ND_constant_color3")
i = n.GetShaderOutput("out")
print("Color3 Sdr type:", i.GetType())
print("Color3 Sdf type:", (str(i.GetTypeAsSdfType()[0]), i.GetTypeAsSdfType()[1]))
n = r.GetNodeByIdentifier("ND_constant_filename")
i = n.GetShaderOutput("out")
print("filename Sdr type:", i.GetType())
print("filename Sdf type:", (str(i.GetTypeAsSdfType()[0]), i.GetTypeAsSdfType()[1]))
n = r.GetNodeByIdentifier("ND_lama_add_edf")
i = n.GetShaderOutput("out")
print("EDF Sdr type:", i.GetType())
print("EDF Sdf type:", (str(i.GetTypeAsSdfType()[0]), i.GetTypeAsSdfType()[1]))
```
Will return:
```
Color3 Sdr type: color
Color3 Sdf type: ('color3f', '')
filename Sdr type: string
filename Sdf type: ('asset', '')
EDF Sdr type: EDF
EDF Sdf type: ('token', 'EDF')
```
So, here, you should do this instead:
```C++
    const auto sdfTypeTuple = fShaderAttributeDef->GetTypeAsSdfType();
    if (sdfTypeTuple.second.empty()) {
        return sdfTypeTuple.first.GetAsToken().GetString();
    } else {
        return sdfTypeTuple.second();
    }
```

Reply:
@stefanminning-autodesk I heard that this metadata might get backported into Ufe v5. If that is true, then this ifdef check (and the others) should be modified to not be version dependent.  See https://github.com/Autodesk/maya-usd/blob/dev/cmake/modules/FindUFE.cmake for other examples of this where we look inside of header files (from Ufe) for a feature and set a flag if we find the new feature. That way if you backport to Ufe v5 no other MayaUsd changes will be required.